src/learning/nnet.rs

```diff
@@ -328,10 +311,9 @@ impl<'a, T: Criterion> BaseNeuralNet<'a, T> {
             }
         }
 
-        let mut grad = Vec::with_capacity(self.layer_sizes.len() - 1);
-        let mut capacity = 0;
+        let mut gradients = Vec::new();
 
-        for (l, activ_item) in activations.iter().enumerate().take(self.layer_sizes.len() - 1) {
+        for (l, activ_item) in activations[..self.layer_sizes.len() - 1].iter().enumerate() {
```

Comment:
Maybe it is just me but personally I prefer `take` to indexing like this. The compiler probably boils them down to the same thing but I worry indexing may add some extra bound checks or something?

(Feel free to make the case for indexing instead!)